File: lib/cli/commands/Refund.ts
import {
  extractClaimPublicKeyFromReverseSwapTree,
  extractClaimPublicKeyFromSwapTree,
} from 'boltz-core';
import { Arguments } from 'yargs';
import { constructRefundTransaction } from '../../Core';
import { stringify } from '../../Utils';
import BuilderComponents from '../BuilderComponents';
import { prepareTx } from '../Command';

export const command =
  'refund <network> <privateKey> <timeoutBlockHeight> <redeemScript> <rawTransaction> <destinationAddress> [feePerVbyte] [blindingKey] [discountCT]';

export const describe = 'refunds submarine or chain to chain swaps';

export const builder = {
  network: BuilderComponents.network,
  privateKey: BuilderComponents.privateKey,
  timeoutBlockHeight: {
    describe: 'timeout block height of the swap',
    type: 'number',
  },
  redeemScript: BuilderComponents.redeemScript,
  rawTransaction: BuilderComponents.rawTransaction,
  destinationAddress: BuilderComponents.destinationAddress,
  feePerVbyte: BuilderComponents.feePerVbyte,
  blindingKey: BuilderComponents.blindingKey,
  discountCT: BuilderComponents.discountCT,
};

export const handler = async (argv: Arguments<any>): Promise<void> => {
  const {
    keys,
    walletStub,
    swapOutput,
    transaction,
    redeemScript,
    destinationAddress,
  } = await prepareTx(argv, [
    extractClaimPublicKeyFromSwapTree,
    extractClaimPublicKeyFromReverseSwapTree,
  ]);

  const refundTransaction = constructRefundTransaction(
    walletStub,
    [
      {
        ...swapOutput,
        keys,
        redeemScript,
        txHash: transaction.getHash(),
      } as any,
    ],
    destinationAddress,
    argv.timeoutBlockHeight,
    argv.feePerVbyte,
  ).toHex();

  console.log(stringify({ refundTransaction }));
};

File: boltzr-cli/src/tx/refund.rs
use crate::tx::utils::construct_transaction;
use anyhow::Result;
use boltz_core::{Network, Transaction, utils::InputType};

#[allow(clippy::too_many_arguments)]
pub fn refund_utxo(
    network: Network,
    timeout_block_height: u32,
    private_key: Vec<u8>,
    swap_tree_or_redeem_script: &str,
    raw_transaction: Vec<u8>,
    destination_address: &str,
    fee_per_vbyte: f64,
    blinding_key: Option<Vec<u8>>,
) -> Result<Transaction> {
    construct_transaction(
        network,
        InputType::Refund(timeout_block_height),
        private_key,
        swap_tree_or_redeem_script,
        raw_transaction,
        destination_address,
        fee_per_vbyte,
        blinding_key,
    )
}
